crates/datasources/src/object_store/mod.rs

```diff
@@ -31,35 +48,275 @@ impl FromStr for FileType {
     type Err = ObjectStoreSourceError;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        let s = s.to_lowercase();
-        match s.as_str() {
-            "parquet" => Ok(Self::Parquet),
-            "csv" => Ok(Self::Csv),
-            "json" => Ok(Self::Json),
-            _ => Err(Self::Err::NotSupportFileType(s)),
+        let typ = if s.eq_ignore_ascii_case("parquet") {
+            Self::Parquet
+        } else if s.eq_ignore_ascii_case("csv") {
+            Self::Csv
+        } else if s.eq_ignore_ascii_case("json") {
+            Self::Json
+        } else {
+            return Err(Self::Err::NotSupportFileType(s.to_owned()));
+        };
+        Ok(typ)
+    }
+}
+
+#[async_trait]
+pub trait FileTypeAccess: Debug + Send + Sync {
+    /// Gets the schema using objects.
+    async fn get_schema(
+        &self,
+        store: Arc<dyn ObjectStore>,
+        objects: &[ObjectMeta],
+    ) -> Result<SchemaRef>;
+
+    /// Transform into an execution plan.
+    async fn get_exec_plan(
+        &self,
+        ctx: &SessionState,
+        store: Arc<dyn ObjectStore>,
+        conf: FileScanConfig,
+        filters: &[Expr],
+        predicate_pushdown: bool,
+    ) -> Result<Arc<dyn ExecutionPlan>>;
+}
+
+#[async_trait]
+pub trait ObjStoreAccess: Debug + Display + Send + Sync {
+    /// Returns the base URL against which the object store is registered.
+    ///
+    /// Examples:
+    /// * `s3//bucket/path/to/file.csv`: `s3://bucket`
+    /// * `/some/local/file`: `file://`
+    fn base_url(&self) -> Result<ObjectStoreUrl>;
```

Comment:
Could we add a quick example for http too since that's a bit special? I think that'll make it clear that `ObjectStoreUrl` here is acting as a key to the registry and not being used to build up the actual url to fetch the object.